### ISSUE TITLE
Refills failed when HPDCACHE_ACCESS_WORDS is 1 or equal to HPDCACHE_CL_WORDS

### ISSUE THREAD

**Issue by AileonN** [openhwgroup/cv-hpdcache#3]
Hello! I am Noelia Oliete. 

During the integration of the HPDC and OpenPiton for the GSoC'23, I detected that the refills don't work when the configurable parameter HPDCACHE_ACCESS_WORDS is equal to 1 or to HPDCACHE_CL_WORDS (Now I am using  16B/cache line). 

- For the first case, HPDCACHE_ACCESS_WORDS=1, the function _hpdcache_compute_data_ram_cs_  (located in _hpdcache_memctrl.sv_) returns the following compilation error: 

`End index -1 of part-select into 'word_i' is out of bounds.`

This error is due to the next line since we cannot index _word_i_ with "[0 +:0]".
`off = hpdcache_uint'(word_i[0 +: $clog2(HPDCACHE_ACCESS_WORDS)]);`

To solve this compilation error, I would like to propose the following code (it works for Verilator and Questa/Modelsim). Notice that when HPDCACHE_ACCESS_WORDS, the new parameter HPDCACHE_ACCESS_WORDS_LOG2 takes the value 1 to avoid a compilation error in Verilator. 

```systemverilog
    function automatic hpdcache_data_row_enable_t hpdcache_compute_data_ram_cs( 
            input hpdcache_req_size_t size_i,
            input hpdcache_word_t     word_i);
        localparam int unsigned HPDCACHE_ACCESS_WORDS_LOG2 = (HPDCACHE_ACCESS_WORDS==1) ? 1 : $clog2(HPDCACHE_ACCESS_WORDS);
        hpdcache_data_row_enable_t ret;
        hpdcache_uint32 off;

        case (size_i)
            3'h0,
            3'h1,
            3'h2,
            3'h3:    ret = hpdcache_data_row_enable_t'({1{1'b1}});
            3'h4:    ret = hpdcache_data_row_enable_t'({2{1'b1}});
            3'h5:    ret = hpdcache_data_row_enable_t'({4{1'b1}});
            default: ret = hpdcache_data_row_enable_t'({8{1'b1}});
        endcase

        off = (HPDCACHE_ACCESS_WORDS==1) ? hpdcache_uint'(0) : hpdcache_uint'(word_i[0 +: HPDCACHE_ACCESS_WORDS_LOG2]);
        return hpdcache_data_row_enable_t'(ret << off);      
    endfunction 
``` 
 

- For the second case, HPDCACHE_ACCESS_WORDS=HPDCACHE_CL_WORDS, the system doesn't work properly since the refill becomes valid before the refill data, like the tag, comes. 

Hope this could help. 
Thank you,
Noelia.

**Comment by cfuguet** [comment]
Hello @AileonN,

Thank you very much for reporting this issue. I confirmed it on my side.

These were part of the configurations that were not yet tested.

I did one bugfix with your suggestion and I completed that bugfix to solve an issue with AMOs when HPDCACHE_ACCESS_WORDS=1.

I also did another bugfix to handle the case where HPDCACHE_ACCESS_WORDS == HPDCACHE_CL_WORDS.

I'm currently running some non-regression tests on those configurations plus a few more to be sure that everything works as expected. I hope I will push soon these fixes.

For the first commit, if you give me your approval, I will put you as co-author.

Thanks again,

Regards,

César
